### PR TITLE
Update source and test

### PR DESCRIPTION
**Pull request by jingyaokang**
Create source.py;
Update test for type generator;

### CODE REVIEW

**Comment by jingyaokang** [comment]
Update test for type kafka

File: python/test/test_source.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+from timeplus import Source, Stream
+
+
+def test_source_generator(test_environment):
+    stream_name = "iot"
+    source_name = "iot"
+    # create a new stream
+    stream = Stream(env=test_environment).name(stream_name)
+
+    try:
```
Comment (gangtao) [inline python/test/test_source.py:11]: one improvement here is that you can call stream.exist() to check if the stream already exist, if yes, delete it and recreate.

File: python/test/test_source.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        .description(source_description)
+        .create()
+    )
+
+    source_list = Source(env=test_environment).list()
+    source_id = source.metadata().id
+    assert source_name in [q.name for q in source_list]
+
+    assert Source(env=test_environment).name(source_name).exist(source_id)
+
```
Comment (gangtao) [inline python/test/test_source.py:44]: before, delete, to make sure source is running, you need to run a query on the stream to valid the source is generating new data on that stream.

File: python/test/test_source.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        Stream(env=test_environment)
+        .name(stream_name)
+        .column("raw", "string")
+        .create()
+    )
+
+    source_type = "kafka"
+    source_stream = "test_kafka"
+    source_properties = {
+    "data_type": "text",
```
Comment (gangtao) [inline python/test/test_source.py]: formating the json here

File: python/test/test_source.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    "offset": "latest",
+    "sasl": "none",
```
Comment (gangtao) [inline python/test/test_source.py:77]: `sasl = None` instead of `sasl = "none"` ?

File: python/timeplus/source.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        Sends a request to the API to create the source.
+
+        Returns:
+        Source: The current source object.
+
+        Raises:
+        ApiException: If an error occurs during the API call.
+        """
+        body = {
+            # "name": self._name,
```
Comment (gangtao) [inline python/timeplus/source.py]: you can remove those comments

File: python/timeplus/source.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            # "type": self._type,
+            # "stream": self._stream,
+            # "properties": self._properties
+        }
+
+        if self._name:
```
Comment (gangtao) [inline python/timeplus/source.py:118]: it is better throw exception in requirement parameter is None

File: python/timeplus/source.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    def metadata(self):
+        """
+        Returns the metadata of the source.
+
+        Returns:
+        Any: The metadata of the source.
+        """
+        return self._metadata
+
+    def update(self, source_id, body):
```
Comment (gangtao) [inline python/timeplus/source.py]: we may need hide the update for now, it is a little bit tricky as some of the property cannot be updated

File: python/timeplus/source.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        source_id (str): The id of the source to check.
+
+        Returns:
+        bool: True if the source exists, False otherwise.
+        """
+        try:
+            self._api_instance.v1beta2_sources_id_get(source_id)
+            return True
+        except ApiException:
+            return False
```
Comment (gangtao) [inline python/timeplus/source.py:234]: please consider add start/stop source
https://docs.timeplus.com/rest.html#tag/Sources-v1beta2/paths/~1v1beta2~1sources~1%7Bid%7D~1start/post
https://docs.timeplus.com/rest.html#tag/Sources-v1beta2/paths/~1v1beta2~1sources~1%7Bid%7D~1stop/post

File: python/timeplus/source.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+        Args:
+        source_id (str): The id of the source to update.
+        body (dict): A dictionary containing the parameters to update.
+
+        Raises:
+        ApiException: If an error occurs during the API call.
+        """
+        self._api_instance.v1beta2_sources_id_put(source_id, body)
+
```
Comment (gangtao) [inline python/timeplus/source.py:222]: we need an id() function to get/set the id

you can refer to this set/get pattern
```python
    
    def id(self, *args):
        if len(args) == 0:
            print("Getting id...")
            return self._metadata.id
        elif len(args) == 1:
            new_value = args[0]
            print("Setting id...")
            self._metadata.id= new_value
        else:
            raise ValueError("value() accepts at most 1 argument")
```

source.id('uid').delete()

**Review by gangtao — CHANGES_REQUESTED** [review]
please check my comment, overall I think it is good

**Review by gangtao — APPROVED** [review]

**Comment by jovezhong** [comment]
(Jove Github Bot) added it to the current sprint.

**Comment by jovezhong** [comment]
(Jove Github Bot) moved this ticket out of the GitHub project(up to 1200 tickets for one project).

**Pull request merged**